### PR TITLE
MON-2552: Add support for user-defined alerting rules

### PR DESCRIPTION
**Pull request by jan--f**
* [ ] TODO: add CHANGELOG entry for this change.

### CODE REVIEW

**Comment by jan--f** [comment]
/label docs-approved
Adding doc approval as per agreement.
Ping @juzhao for qe approval.
Ping @Senthamilarasu-STA for px approval.

**Review by JoaoBraveCoding — APPROVED** [review]
Small nits but LGTM! Since it's still not an official feature not sure if we should add tests or we usually do that later down the line

File: pkg/alert/rule_controller.go [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+
+	<-ctx.Done()
+}
+
+// keyFunc derives a queue or cache key for the given object, while properly
+// handling tombstone objects.
+func (rc *RuleController) keyFunc(obj interface{}) (string, bool) {
+	k, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Errorf("Creating AlertingRule key failed: %v", err)
```
[the remark this answers is not in the document]
Comment (jan--f) [reply pkg/alert/rule_controller.go]: I think this message is actually correct, its about the **key** of the `AlertingRule` object. But I agree that the message can be improved, pushed a proposal.

File: pkg/alert/rule_controller.go [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+	if !ok {
+		return
+	}
+
+	klog.V(4).Infof("AlertingRule deleted: %s", key)
+	rc.enqueue(key)
+}
+
+// handleAlertingRuleUpdate handles update events for the AlertingRule informer.
+func (rc *RuleController) handleAlertingRuleUpdate(oldObj, newObj interface{}) {
```
[the remark this answers is not in the document]
Comment (jan--f) [reply pkg/alert/rule_controller.go:177]: Agreed, done.

**Comment by jan--f** [comment]
> Since it's still not an official feature not sure if we should add tests or we usually do that later down the line

I agree 100% on the tests, I don't have the bandwidth to write those before my break and I would like to get this in 4.11. Unless someone has objections or a patch, I'd be ok adding this without e2e tests.

**Review by simonpasquier — COMMENTED** [review]
/lgtm

**Comment by juzhao** [comment]
@jan--f @simonpasquier 
tested with PR
launch 4.11.0-0.ci-2022-06-17-054342,openshift/cluster-monitoring-operator#1675
enabled UWM and create AlertingRule
```
# oc create -f - << EOF
apiVersion: monitoring.openshift.io/v1alpha1
kind: AlertingRule
metadata:
  name: example
  namespace: test
spec:
  groups:
  - name: example-rules
    rules:
    - alert: ExampleAlert
      expr: vector(1)
EOF
error: resource mapping not found for name: "example" namespace: "test" from "STDIN": no matches for kind "AlertingRule" in version "monitoring.openshift.io/v1alpha1"
ensure CRDs are installed first
```

could see 0000_50_cluster-monitoring-operator_00_0alertingrule-config-custom-resource-definition.yaml in manifests
```
# oc -n openshift-monitoring exec -c cluster-monitoring-operator cluster-monitoring-operator-6db7bc969c-jc4pr -- cat manifests/0000_50_cluster-monitoring-operator_00_0alertingrule-config-custom-resource-definition.yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  annotations:
    api-approved.openshift.io: https://github.com/openshift/api/pull/1179
    release.openshift.io/feature-gate: TechPreviewNoUpgrade
    description: OpenShift Monitoring alerting rules
  name: alertingrules.monitoring.openshift.io
spec:
  group: monitoring.openshift.io
  names:
    kind: AlertingRule
    listKind: AlertingRuleList
    plural: alertingrules
    singular: alertingrule
...

no alertingrules crd
```
# oc get crd | grep -i monitoring
alertmanagerconfigs.monitoring.coreos.com                         2022-06-17T09:36:39Z
alertmanagers.monitoring.coreos.com                               2022-06-17T09:36:42Z
podmonitors.monitoring.coreos.com                                 2022-06-17T09:36:44Z
probes.monitoring.coreos.com                                      2022-06-17T09:36:46Z
prometheuses.monitoring.coreos.com                                2022-06-17T09:36:49Z
prometheusrules.monitoring.coreos.com                             2022-06-17T09:36:52Z
servicemonitors.monitoring.coreos.com                             2022-06-17T09:36:54Z
thanosrulers.monitoring.coreos.com                                2022-06-17T09:36:57Z
```
is the issue related to this PR or https://github.com/openshift/api/pull/1179

**Comment by juzhao** [comment]
bug https://bugzilla.redhat.com/show_bug.cgi?id=2099513, not sure if it's expected

**Comment by juzhao** [comment]
since https://bugzilla.redhat.com/show_bug.cgi?id=2099513 is expected
add /label qe-approved

**Pull request merged**